pkg/etcdmgr/bootstrap/discovery.go

```diff
@@ -18,49 +18,71 @@ package bootstrap
 import (
 	"encoding/json"
 	"fmt"
-	"io/ioutil"
 	"net/http"
 	"strconv"
 	"strings"
+	"time"
 
 	"github.com/coreos/etcd/client"
 	"github.com/coreos/etcd/store"
 	"golang.org/x/net/context"
 	"golang.org/x/net/context/ctxhttp"
 )
 
-func isQuorumFull(token string) (bool, error, []string) {
-	res, err := queryDiscoveryService(token + "/_config/size")
+const (
+	discoveryRetryAttempts = 10
+)
+
+func isQuorumFull(token string) (bool, []string, error) {
+	size, err := getClusterSize(token)
 	if err != nil {
-		return false, fmt.Errorf("cannot get discovery url cluster size: %v", err), []string{}
+		return false, nil, err
 	}
+	logger.Infof("cluster size is: %d", size)
 
-	size, _ := strconv.ParseInt(*res.Node.Value, 10, 16)
-	clusterSize := int(size)
-	logger.Infof("cluster max size is: %d", clusterSize)
-
-	currentNodes, _ := GetCurrentNodesFromDiscovery(token)
+	currentNodes, _ := GetCurrentNodesFromDiscovery(token, size)
 	logger.Infof("currentNodes: %+v", currentNodes)
-	if len(currentNodes) < clusterSize {
-		return false, nil, []string{}
+	if len(currentNodes) < size {
+		return false, []string{}, nil
 	}
-	return true, nil, currentNodes
+	return true, currentNodes, nil
 
 }
 
-// queryDiscoveryService reads a key from a discovery url.
-func queryDiscoveryService(token string) (*store.Event, error) {
-	ctx, _ := context.WithTimeout(context.Background(), DefaultClientTimeout)
-	resp, err := ctxhttp.Get(ctx, http.DefaultClient, token)
+func getClusterSize(token string) (int, error) {
+	res, err := queryDiscoveryService(token + "/_config/size")
 	if err != nil {
-		return nil, err
+		return -1, fmt.Errorf("cannot get discovery url cluster size: %v", err)
 	}
-	defer func() { _ = resp.Body.Close() }()
-	if resp.StatusCode != http.StatusOK {
-		body, _ := ioutil.ReadAll(resp.Body)
-		return nil, fmt.Errorf("status code %d from %q: %s", resp.StatusCode, token, body)
+
+	size, _ := strconv.ParseInt(*res.Node.Value, 10, 16)
+	return int(size), nil
+}
+
+// queryDiscoveryService reads a key from a discovery url.
+func queryDiscoveryService(token string) (*store.Event, error) {
+	var resp *http.Response
+	var err error
+
+	// retry the http request in case of network errors
+	for i := 1; i <= discoveryRetryAttempts; i++ {
+		ctx, _ := context.WithTimeout(context.Background(), DefaultClientTimeout)
+		resp, err = ctxhttp.Get(ctx, http.DefaultClient, token)
+		if err == nil && resp.StatusCode == http.StatusOK {
+			break
+		}
+
+		resp.Body.Close()
```

Comment:
i recall being bit in the past by a HTTP response body being nil and then the Close() causing a nil ref.  I think that was with the HTTP communication with elastic search queries in global way back, but my memory is fuzzy.

consider guarding here a bit against resp.Body being nil, and also in the defer statement below that does the same thing.  I'm not super positive about this guidance, but my fuzzy memory is clear enough that being careful here might be wise.

EDIT: actually, ctxhttp may have entirely different guarantees about the response body than the standard net/http package does.  you may know better than i do.